packages/bot-analytics/test/unit/volume.unit.test.ts

```diff
@@ -0,0 +1,652 @@
+import { inititalizeChains } from "../../src/db/init";
+import assert from "assert";
+import { handleRange } from "../../src/analytics";
+import { generateBlockHeaderToDbBlock } from "../../src/util/util";
+import { Sdk } from "../../.graphclient";
+
+describe("Volume tracking", () => {
+  let prisma: PrismaClient | undefined;
+  let stop: () => Promise<void> | undefined;
+  let resetDb: () => Promise<void> | undefined;
+
+  const chainId = 0;
+
+  const chain: Chain = {
+    name: "test",
+    chainId,
+  };
+
+  const blocks: Record<string, BlockHeader> = {
+    "1": {
+      number: 1,
+      hash: "0x1",
+      timestamp: Date.now() / 1000,
+    } as BlockHeader,
+    "2": {
+      number: 2,
+      hash: "0x2",
+      timestamp: Date.now() / 1000 + 2000,
+    } as BlockHeader,
+    "3": {
+      number: 3,
+      hash: "0x3",
+      timestamp: Date.now() / 1000 + 2000 * 2,
+    } as BlockHeader,
+  };
+
+  const context: ChainContext = {
+    ...chain,
+    blockFinality: 0,
+    getBlock: async (number: string | number) => {
+      return blocks[number];
+    },
+    multicall2: {} as any,
+    subgraphMaxFirstValue: 100,
+    everyXBlock: 1,
+  };
+
+  const blockHeaderToDbBlock = generateBlockHeaderToDbBlock(context);
+
+  const account0: Account = {
+    address: "account0",
+  };
+
+  const account1: Account = {
+    address: "account1",
+  };
+
+  const account3: Account = {
+    address: "account3",
+  };
+
+  const token0: TokenWithoutId = {
+    address: "token0",
+    symbol: "tkn0",
+    decimals: 18,
+    chainId,
+  };
+
+  const token1: TokenWithoutId = {
+    address: "token1",
+    symbol: "tkn1",
+    decimals: 18,
+    chainId,
+  };
+
+  const tokens = {
+    [token0.address]: token0,
+    [token1.address]: token1,
+  };
+
+  const getOrCreateTokenFn = async (prisma: PrismaTx, address: string) => {
+    let token = await prisma.token.findFirst({
+      where: {
+        address,
+        chainId: chainId,
+      },
+    });
+
+    if (!token) {
+      token = await prisma.token.create({
+        data: tokens[address],
+      });
+    }
+    return token;
+  };
+
+  const generateMockSdk = (results: Record<string, GetVolumesResult[]>) =>
+    ({
+      getVolumes: async (
+        params: GetParamsVolumes
+      ): Promise<GetVolumeResults> => {
+        const _results = results[params.currentBlockNumber];
+        return {
+          accountVolumeByPairs: _results.slice(
+            params.skip,
+            params.skip === 0 ? params.first : params.first + params.skip
+          ),
+        };
+      },
+    } as Sdk);
+
+  const getLatestActivityWithAddressAndType = async (
+    address: string,
+    asMaker: boolean
+  ) => {
+    return prisma!.accountActivity.findFirst({
+      orderBy: {
+        id: "desc",
+      },
+      where: {
+        accountId: address,
+        asMaker: asMaker,
+      },
+    });
+  };
+
+  before(async () => {
+    const result = await startDb();
```

Comment:
Not sure i would deem this a unit test, when you have to start a database up for it to run

Reply:
Yes this was the issue I told you during our call, the code is so tightly bounded with DB that I'm afraid unit testing is hard. We could build a fully mock db but it will require so much time abstracting this logic, I'm unsure it's needed

Reply:
Agreed, lets not do more for now. But we should keep this in mind when we expand on the analytics, and see if we cannot write the code so that we it is not to db dependent.